### ISSUE TITLE
ScalaReflectionException in TypeBuilder

### ISSUE THREAD

**Issue by JanGe** [http4s/rho#203]
I'm trying to define an endpoint in Rho that gets passed an entity in JSON Patch format by using the definitions from [diffson-circe](https://github.com/gnieh/diffson/blob/master/circe/src/main/scala/gnieh/diffson/CirceInstance.scala):

```scala
import io.circe.generic.auto._
import org.http4s.circe._
import org.http4s.rho._
import org.http4s.rho.swagger._

object Service {

  import gnieh.diffson.circe._
  import gnieh.diffson.circe.DiffsonProtocol._

  val service: RhoService = new RhoService {
    PATCH / "hello" ^ jsonOf[JsonPatch] |>> { _: JsonPatch =>
      Ok("Hello World.")
    }
  }
}
```

When hitting the `GET /swagger.json` endpoint, the `JsonPatch` definition is missing from the spec and an exception is logged:
```
[scala-execution-context-global-63] INFO  o.h.r.s.TypeBuilder - Failed to build model for type JsonPatchSupport.this.Test
scala.ScalaReflectionException: type JsValue is not a class
	at scala.reflect.api.Symbols$SymbolApi.asClass(Symbols.scala:272)
	at scala.reflect.api.Symbols$SymbolApi.asClass$(Symbols.scala:272)
	at scala.reflect.internal.Symbols$SymbolContextApiImpl.asClass(Symbols.scala:94)
	at org.http4s.rho.swagger.TypeBuilder$.org$http4s$rho$swagger$TypeBuilder$$isObjectEnum(TypeBuilder.scala:144)
	at org.http4s.rho.swagger.TypeBuilder$DataType$.fromType(TypeBuilder.scala:291)
	at org.http4s.rho.swagger.TypeBuilder$.$anonfun$typeToProperty$1(TypeBuilder.scala:203)
	at scala.PartialFunction.applyOrElse(PartialFunction.scala:123)
	at scala.PartialFunction.applyOrElse$(PartialFunction.scala:122)
	at scala.PartialFunction$$anon$1.applyOrElse(PartialFunction.scala:252)
	at org.http4s.rho.swagger.TypeBuilder$.typeToProperty(TypeBuilder.scala:185)
	at org.http4s.rho.swagger.TypeBuilder$.paramSymToProp(TypeBuilder.scala:179)
	at org.http4s.rho.swagger.TypeBuilder$.$anonfun$modelToSwagger$1(TypeBuilder.scala:160)
	at scala.collection.immutable.List.map(List.scala:287)
	at org.http4s.rho.swagger.TypeBuilder$.modelToSwagger(TypeBuilder.scala:160)
	at org.http4s.rho.swagger.TypeBuilder$.go$1(TypeBuilder.scala:70)
	at org.http4s.rho.swagger.TypeBuilder$.$anonfun$collectModels$9(TypeBuilder.scala:94)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:241)
	at scala.collection.immutable.HashSet$HashSet1.foreach(HashSet.scala:320)
	at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:976)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:241)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:238)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:104)
	at org.http4s.rho.swagger.TypeBuilder$.$anonfun$collectModels$8(TypeBuilder.scala:93)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:241)
	at scala.collection.immutable.Set$Set1.foreach(Set.scala:95)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:241)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:238)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:104)
	at org.http4s.rho.swagger.TypeBuilder$.go$1(TypeBuilder.scala:90)
	at org.http4s.rho.swagger.TypeBuilder$.$anonfun$collectModels$6(TypeBuilder.scala:80)
	at scala.collection.immutable.List.flatMap(List.scala:335)
	at org.http4s.rho.swagger.TypeBuilder$.go$1(TypeBuilder.scala:74)
	at org.http4s.rho.swagger.TypeBuilder$.collectModels(TypeBuilder.scala:103)
	at org.http4s.rho.swagger.TypeBuilder$.collectModels(TypeBuilder.scala:23)
	at org.http4s.rho.swagger.SwaggerModelsBuilder.$anonfun$collectDefinitions$1(SwaggerModelsBuilder.scala:51)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1(TraversableOnce.scala:157)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(TraversableOnce.scala:157)
	at scala.collection.immutable.Set$Set2.foreach(Set.scala:130)
	at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:157)
	at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:155)
	at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
	at org.http4s.rho.swagger.SwaggerModelsBuilder.collectDefinitions(SwaggerModelsBuilder.scala:51)
	at org.http4s.rho.swagger.SwaggerModelsBuilder.mkSwagger(SwaggerModelsBuilder.scala:25)
	at org.http4s.rho.swagger.SwaggerSupport$.$anonfun$createSwagger$1(SwaggerSupport.scala:44)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1(TraversableOnce.scala:157)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(TraversableOnce.scala:157)
	at scala.collection.Iterator.foreach(Iterator.scala:929)
	at scala.collection.Iterator.foreach$(Iterator.scala:929)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
	at scala.collection.IterableLike.foreach(IterableLike.scala:71)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:157)
	at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:155)
	at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
	at org.http4s.rho.swagger.SwaggerSupport$.createSwagger(SwaggerSupport.scala:44)
	at org.http4s.rho.swagger.SwaggerSupport$.swaggerSpec$lzycompute$1(SwaggerSupport.scala:27)
	at org.http4s.rho.swagger.SwaggerSupport$.swaggerSpec$1(SwaggerSupport.scala:25)
	at org.http4s.rho.swagger.SwaggerSupport$.$anonfun$apply$2(SwaggerSupport.scala:31)
	at org.http4s.rho.swagger.SwaggerSupport$$anon$1.response$lzycompute(SwaggerSupport.scala:58)
	at org.http4s.rho.swagger.SwaggerSupport$$anon$1.response(SwaggerSupport.scala:55)
	at org.http4s.rho.swagger.SwaggerSupport$$anon$1.$anonfun$new$1(SwaggerSupport.scala:61)
	at shapeless.ops.FnToProductInstances$$anon$1.$anonfun$apply$1(fntoproduct.scala:37)
	at org.http4s.rho.bits.HListToFunc$$anon$2.$anonfun$conv$2(HListToFunc.scala:35)
	at org.http4s.rho.bits.PathTree$.$anonfun$makeLeaf$2(PathTree.scala:85)
	at org.http4s.rho.bits.ResultResponse.map(ResultResponse.scala:34)
	at org.http4s.rho.bits.ResultResponse.map$(ResultResponse.scala:33)
	at org.http4s.rho.bits.SuccessResponse.map(ResultResponse.scala:50)
	at org.http4s.rho.bits.PathTree$.$anonfun$makeLeaf$1(PathTree.scala:84)
	at org.http4s.rho.bits.PathTree$SingleLeaf.attempt(PathTree.scala:125)
	at org.http4s.rho.bits.PathTree$Node.walk(PathTree.scala:247)
	at org.http4s.rho.bits.PathTree$Node.walk$(PathTree.scala:212)
	at org.http4s.rho.bits.PathTree$MatchNode.walk(PathTree.scala:264)
	at org.http4s.rho.bits.PathTree$Node.walk(PathTree.scala:222)
	at org.http4s.rho.bits.PathTree$Node.walk$(PathTree.scala:212)
	at org.http4s.rho.bits.PathTree$MatchNode.walk(PathTree.scala:264)
	at org.http4s.rho.bits.PathTree$Node.walkTree(PathTree.scala:208)
	at org.http4s.rho.bits.PathTree$Node.walkTree$(PathTree.scala:206)
	at org.http4s.rho.bits.PathTree$MatchNode.walkTree(PathTree.scala:264)
	at org.http4s.rho.bits.PathTree.getResult(PathTree.scala:45)
	at org.http4s.rho.CompileService$.$anonfun$foldServices$2(CompileService.scala:44)
	at scala.Function1.$anonfun$andThen$1(Function1.scala:52)
	at scala.Function1.$anonfun$andThen$1(Function1.scala:52)
	at cats.data.KleisliSemigroup.$anonfun$combine$1(Kleisli.scala:254)
	at cats.data.Kleisli.apply(Kleisli.scala:63)
	at org.http4s.server.Router$.$anonfun$define$3(Router.scala:35)
	at org.http4s.server.blaze.Http1ServerStage.liftedTree1$1(Http1ServerStage.scala:107)
	at org.http4s.server.blaze.Http1ServerStage.$anonfun$runRequest$2(Http1ServerStage.scala:107)
	at fs2.Strategy$$anon$3$$anon$4.run(Strategy.scala:54)
	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:140)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Comment by bryce-anderson** [comment]
Looks like a bug, thanks for the report! I don't think I'll be able to take
care of it for a few days, but if you're interested in taking a swing at
it, I could probably provide some guidance.

**Comment by zcox** [comment]
I see a very similar stack trace when using Shapeless `Coproduct`:

```
2018-06-25 16:39:05.117 [scala-execution-context-global-23] INFO  org.http4s.rho.swagger.TypeBuilder - Failed to build model for type shapeless.Inl[H,T]
scala.ScalaReflectionException: type H is not a class
```

I wrote a [small example that reproduces the stack trace](https://github.com/zcox/rho-coproduct-test). It seems to work very well with a sealed trait hierarchy, but not with Coproduct.

**Comment by nightscape** [comment]
I just ran into the same problem. It seems to me that the `isObjectEnum` method is missing the `sym.isClass` check that most other methods have:
https://github.com/http4s/rho/blob/d05998fdbd2f15df0b18bdd7f30437db02c766f0/swagger/src/main/scala/org/http4s/rho/swagger/TypeBuilder.scala#L138
@reactormonk @bryce-anderson I saw this was introduced in [this commit](https://github.com/http4s/rho/commit/a54c6353495eb07b82160f0def6e3371580b52ef). I know it's been a long time, but do you happen to remember if this on purpose?

**Comment by RafalSumislawski** [comment]
I've managed to implement tests reproducing issues reported by @JanGe and @zcox . Basing on these tests I can confirm that the direct cause for scala.ScalaReflectionException being thrown is the lack of `sym.isClass` in  `isObjectEnum` as @nightscape pointed out. Having already worked a bit with the `TypeBuilder` I see no reason for omitting the `sym.isClass`. Adding it doesn't break any tests.

The bad news is that underneath this simple bug, hides the fact that `TypeBuilder` is incapable of building a correct model for `JsonPatch` or `Coproduct`. In both cases `TypeBuilder` fails to resolve type parameters to concrete types. In case of `JsonPatch` the problem lies in inner classes like `gnieh.diffson.JsonPatchSupport.Test` using the type parameter `JsValue` of their outer trait.
In case of `Coproduct` well... it's shapeless, type parameters can't get much more complex that that, can they?

I haven't found any way to solve the `JsonPatch` issue. The `Coproduct` is solvable by treating it as a special case. I'm not sure what the maintainers of rho will say to that.

For now, I will PR the unit tests and the `sym.isClass`.

**Comment by zarthross** [comment]
I've added issue #283 to address the special case for Shapeless `Coproduct`.  That seems the best course of action for that.  

@JanGe Version [0.19.0-M5](https://github.com/http4s/rho/releases/tag/v0.19.0-M5) should have a fix for this issue (see PR #281), can you confirm (once I get it released that is)?

Thanks @RafalSumislawski for running this down and submitting a PR!!!
